packages/kbn-data-stream-adapter/src/create_or_update_component_template.ts

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type {
+  ClusterPutComponentTemplateRequest,
+  IndicesGetIndexTemplateIndexTemplateItem,
+} from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+import type { Logger, ElasticsearchClient } from '@kbn/core/server';
+import { asyncForEach } from '@kbn/std';
+import { retryTransientEsErrors } from './retry_transient_es_errors';
+
+interface CreateOrUpdateComponentTemplateOpts {
+  logger: Logger;
+  esClient: ElasticsearchClient;
+  template: ClusterPutComponentTemplateRequest;
+  totalFieldsLimit: number;
+}
+
+const putIndexTemplateTotalFieldsLimitUsingComponentTemplate = async (
+  esClient: ElasticsearchClient,
+  componentTemplateName: string,
+  totalFieldsLimit: number,
+  logger: Logger
+) => {
+  // Get all index templates and filter down to just the ones referencing this component template
+  const { index_templates: indexTemplates } = await retryTransientEsErrors(
+    () => esClient.indices.getIndexTemplate(),
+    { logger }
+  );
+  const indexTemplatesUsingComponentTemplate = (indexTemplates ?? []).filter(
+    (indexTemplate: IndicesGetIndexTemplateIndexTemplateItem) =>
+      (indexTemplate.index_template?.composed_of ?? []).includes(componentTemplateName)
+  );
+
+  await asyncForEach(
+    indexTemplatesUsingComponentTemplate,
+    async (template: IndicesGetIndexTemplateIndexTemplateItem) => {
+      await retryTransientEsErrors(
+        () =>
+          esClient.indices.putIndexTemplate({
+            name: template.name,
+            body: {
+              ...template.index_template,
+              template: {
+                ...template.index_template.template,
+                settings: {
+                  ...template.index_template.template?.settings,
+                  'index.mapping.total_fields.limit': totalFieldsLimit,
+                },
+              },
+            },
+          }),
+        { logger }
+      );
+    }
+  );
+};
+
+const createOrUpdateComponentTemplateHelper = async (
+  esClient: ElasticsearchClient,
+  template: ClusterPutComponentTemplateRequest,
+  totalFieldsLimit: number,
+  logger: Logger
+) => {
+  try {
+    await retryTransientEsErrors(() => esClient.cluster.putComponentTemplate(template), { logger });
+  } catch (error) {
+    const reason = error?.meta?.body?.error?.caused_by?.caused_by?.caused_by?.reason;
```

Comment:
`error?.meta?.body?.error?.caused_by?.caused_by?.caused_by?.reason;` reads like it's automated [five whys](https://en.wikipedia.org/wiki/Five_whys) 😂

Reply:
😂 yeah that part was a blind copy/pasta from the implementation of the alerts plugin:  https://github.com/elastic/kibana/blob/6c36503a6362435f17e7f2f3bdc0f69974d1e4c5/x-pack/plugins/alerting/server/alerts_service/lib/create_or_update_component_template.ts#L79

But I just tested that error and what we receive is:
```
{
  name: 'ResponseError',
  message: 'illegal_argument_exception\n' +
    '\tRoot causes:\n' +
    '\t\tillegal_argument_exception: Limit of total fields [1] has been exceeded'
}
```

I updated the code. Thanks!